### ISSUE TITLE
Evaluate MeZO Edge Pruning

### ISSUE THREAD

**Issue by MustafaKhan0** [alansun17904/MAAA#8]
### Summary

Run a suite of benchmarks (GP, GT, IOI, Boolean Exp) on the MeZO edge pruning method, to compare accuracy, runtime, and vram usage. 

### Method

- Use implementation from #7 
- Use the same sparsity / hyper parameters as edge-pruning study
- Run benchmarks (gp, gt, ioi, boolean expressions) on ours and standard edge-pruning
- Run both on A100, and record runtime, vram usage, and accuracy
  - Use GPUstat or eyeballing with `nvidia-smi` for vram usage

### Computation

A100 (probably) for runtime evals

### Hypothesis

- Our memory requirements are significantly less than edge pruning
- Our runtime is more than edge pruning
- Final circuit should be very similar / the same

### Acceptance Criteria

- We have experimental results from some / most of the benchmarks that edge pruning used
- Includes memory, runtime, and accuracy metrics

**Comment by alansun17904** [comment]
Setting up pytest with parallel directories: https://stackoverflow.com/questions/41748464/pytest-cannot-import-module-while-python-can

**Comment by MustafaKhan0** [comment]
Exactly what hyper parameters / sparsities should we use? It looks like running both Edge Pruning and MeZO across all values will be fairly resource intensive - should discuss at next meeting.

**Comment by MustafaKhan0** [comment]
One change I made was correctly implementing MeZO using different learning rates for the mask and lambda values - previously, it would just sample from the main edges learning rate for the mask. One reason I made this change was bc the structured pruning paper which edge pruning relies on, notes that the mask and the lambda values should have different learning rates. 

I also made sure to integrate both the node and edge loss, as even though the script as the disable node loss tag, the node loss  is still being used in the compute loss function, and thus should also be optimized. 

I also noticed that as the LR would increase, the loss would increase. However, at the same time, the model sparsity would also increase, which is needed - using the original hyper parameters, the model sparsity is very low (~15% if I remember correctly)

One thing I'm confused about is the optimization of the Lambda parameters. EdgePruning and the structured pruning paper whose method it uses (Wang et al., 2020), note that Lambda is optimized for maximization, creating an adversarial relationship between Lambda and the mask, with respect to the loss. What I don't understand is why said loss would ever decrease. 

Function for the part of the loss which enforces sparsity. This value is added to the KL divergence loss: $$\mathcal L_s = \lambda_1 \cdot (t-s) + \lambda_2 \cdot (t-s)^2$$
Where $$s$$ is the current model sparsity, and $$t$$ is the target model sparsity (which warms-up over a specified number of steps). 

If lambda can be optimized, then it would seem to me that lambda could just continue to increase, compensating for any change in model sparsity, and the loss would either not decrease or continue to increase. That's just my intuition, and is definitely going wrong somewhere or missing something, because the experimental graphs don't match that (at least for backprop). On the other hand, MeZO seems to match this intuition (which makes me think it's some sort of disconnect / flaw). 

**Graphs of Backprop:**

Model (Edge) Sparsity:

<img width="419" height="285" alt="Image" src="https://github.com/user-attachments/assets/74f5e1ea-f664-4396-a2ba-8dbf245b5917" />

Lambda (edge) 1:

<img width="413" height="285" alt="Image" src="https://github.com/user-attachments/assets/1de2c892-3a34-4766-b712-a837ff3db999" />

Loss:

<img width="415" height="294" alt="Image" src="https://github.com/user-attachments/assets/e09ee0d0-f9b0-4529-b33a-f984e3a9fd8f" />


**Graphs of MeZO**

Model (Edge) Sparsity:

<img width="421" height="289" alt="Image" src="https://github.com/user-attachments/assets/a4ce9191-30ac-437e-a0ae-36c4541b74d5" />

Lambda (edge) 1:

<img width="414" height="284" alt="Image" src="https://github.com/user-attachments/assets/955cd28f-7fe6-460a-bdbc-39be716d7c84" />

Loss:

<img width="428" height="279" alt="Image" src="https://github.com/user-attachments/assets/3359ac6e-9be1-44fb-9c62-8dc3474889eb" />

One apparent difference between the implementations of the two methods of optimization is the use of AdamW - backprop uses it, but MeZO does not - just an LR scheduler which increases then decreases linearly. I'm not sure how this would cause such a difference, but it is a difference in implementation. 

Another symptom to note, seen in the graph, is that the loss doesn't seem to be decreasing, and the model doesn't get any more sparse after a certain point is reached in training. I'm not sure why this is. 

While looking into MeZO more, I found that in the intuitive [explanation](https://princeton-nlp.github.io/mezo/) of why MeZO works, the authors mention that MeZO only works well on fine-tuning applications, because the changes / model updates are comparatively low-dimension, compared to pre-training, where updates are needed over almost the entire model. This could prove to be an issue, as the mask is currently randomly initialized - and more comparable to pertaining, rather than fine-tuning. The warm start would likely help with this, but might still be challenging, because even with the warm start, the softened discrete values might not convey enough information to be a helpful prior.

**Comment by MustafaKhan0** [comment]
## Experiment Update

Following the Thursday (July 24) meeting, I made two changes to the experimental setup: 
1. I made the learning rate for all parameters constant (without a warm-up). 
2. I used different, reduced learning rates, and altered the relative difference between the LRs for the mask and lambda. Namely, the LR for the mask being between 1e-3 and 1e-5, and lambda being anywhere from 0 to 3 orders of magnitude smaller.

After just a few experiments, a subset of the parameter sweep we originally planned, I found that these learning rates seemed to be pretty ineffective. Specifically, the model sparsity did not change at all, from initialization. Another anomaly was that the KL loss went negative somehow? To ensure my code changes didn't break anything, I reproduced the general trend we previously had, with larger learning rates.

## Graphs:

**Mask LR: 1e-3, Lambda LR: 1e-3**:

Loss:

<img width="417" height="288" alt="Image" src="https://github.com/user-attachments/assets/f535642f-b40c-4ff1-8373-60e44a247bc1" />

Lambda (edge) 1:

<img width="419" height="291" alt="Image" src="https://github.com/user-attachments/assets/94e75ea1-42c1-4e18-b271-a96cfa7a90c7" />

Lambda (edge) 2:

<img width="414" height="288" alt="Image" src="https://github.com/user-attachments/assets/f288885f-72b4-445a-b618-9cf06d55a760" />

Model (edge) sparsity:

<img width="416" height="292" alt="Image" src="https://github.com/user-attachments/assets/fea71440-a252-40d6-8283-dc2b043857dc" />

KL Loss: 

<img width="415" height="288" alt="Image" src="https://github.com/user-attachments/assets/d00ec091-ea01-4126-aab8-4247838a2fdc" />

**Mask LR: 1e-3, Lambda LR: 1e-4**:

Loss:

<img width="413" height="284" alt="Image" src="https://github.com/user-attachments/assets/a5887184-7e9d-46a5-a071-44c5f70a6861" />

Lambda (edge) 1:

<img width="412" height="283" alt="Image" src="https://github.com/user-attachments/assets/eb43518c-5aa9-4b0a-9366-376cbab14172" />

Lambda (edge) 2:

<img width="423" height="277" alt="Image" src="https://github.com/user-attachments/assets/bd8432b3-0f4b-4b49-9aca-602e27c8e0f9" />

Model (edge) sparsity:

<img width="418" height="287" alt="Image" src="https://github.com/user-attachments/assets/0529380c-34f8-4412-8c95-5c34c6d79cd6" />

KL Loss: 

<img width="425" height="287" alt="Image" src="https://github.com/user-attachments/assets/91c0973f-ac2f-4219-b076-d4f60f431b28" />

**Mask LR: 1e-4, Lambda LR: 1e-5**:

Loss:

<img width="413" height="280" alt="Image" src="https://github.com/user-attachments/assets/87e4ca03-215a-4299-9b37-e4c506883f6e" />

Lambda (edge) 1:

<img width="418" height="286" alt="Image" src="https://github.com/user-attachments/assets/d769e48d-d99f-42b1-9a5e-4706356a2dba" />

Lambda (edge) 2:

<img width="410" height="284" alt="Image" src="https://github.com/user-attachments/assets/f7cfc76b-a503-4f1d-a96b-cad1064f0031" />

Model (edge) sparsity:

<img width="411" height="294" alt="Image" src="https://github.com/user-attachments/assets/7ed9a621-5369-447b-ad24-71ec33333c57" />

KL Loss: 

<img width="423" height="282" alt="Image" src="https://github.com/user-attachments/assets/243a6d12-a8a9-495a-95d9-53dbd71317bf" />

**Mask LR: 1e-1, Lambda LR: 1e-2**:

Loss:

<img width="422" height="289" alt="Image" src="https://github.com/user-attachments/assets/8429b570-17bb-45e5-b380-b86f5f4352b1" />

Lambda (edge) 1:

<img width="414" height="285" alt="Image" src="https://github.com/user-attachments/assets/4092432f-97e2-4352-8ebd-748629b2b9c6" />

Lambda (edge) 2:

<img width="422" height="292" alt="Image" src="https://github.com/user-attachments/assets/2dbd9fb7-c8ee-4947-b754-388fe26f5d46" />

Model (edge) sparsity:

<img width="416" height="282" alt="Image" src="https://github.com/user-attachments/assets/8ab2951b-847e-4f3a-b2c3-746f6775e088" />

KL Loss: 

<img width="410" height="286" alt="Image" src="https://github.com/user-attachments/assets/0290d61e-1487-48ab-b8fc-b7555419deae" />
